.github/workflows/source-and-docs-release.yml

```diff
@@ -0,0 +1,100 @@
+
+name: "Build Python source and docs artifacts"
+
+jobs:
+  source-and-docs:
+    runs-on: ubuntu-22.04
+    steps:
+      - name: "Checkout python/release-tools"
+        uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
+
+      - name: "Checkout ${{ inputs.git_remote }}/cpython"
+        uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
+        with:
+          repository: "${{ inputs.git_remote }}/cpython"
+          ref: "v${{ inputs.cpython_release }}"
+          path: "cpython"
+
+      - name: "Verify CPython commit matches tag"
+        run: |
+          if [[ "${{ inputs.git_commit }}" != "$(cd cpython && git rev-parse HEAD)" ]]; then
+            echo "expected git commit ('${{ inputs.git_commit }}') didn't match tagged commit ('$(git rev-parse HEAD)')"
+            exit 1
+          fi
+
+      - name: "Setup Python"
+        uses: actions/setup-python@65d7f2d534ac1bc67fcd62888c5f4f3d2cb2b236 # v4.7.1
+        with:
+          python-version: 3.11
+
+      - name: "Install source dependencies"
+        run: |
+          python -m pip install --no-deps \
+            -r requirements.txt
+
+      - name: "Install docs dependencies"
+        # Docs aren't built for alpha or beta releases.
+        if: ${{ !(contains(inputs.cpython_release, 'a') || contains(inputs.cpython_release, 'b')) }}
+        run: |
+          python -m pip install \
+            -r cpython/Doc/requirements.txt
+
+          sudo apt-get update
+          sudo apt-get install --yes --no-install-recommends \
+            latexmk texlive-xetex xindy texinfo texlive-latex-base \
+            texlive-fonts-recommended texlive-fonts-extra \
+            texlive-full
+
+      - name: "Build Python release artifacts"
+        run: |
+          cd cpython
+          python ../release.py --export ${{ inputs.cpython_release }}
+
+      - name: "Test Python source tarballs"
+        run: |
+          mkdir -p ./tmp/installation/
+          cp cpython/${{ inputs.cpython_release }}/src/Python-${{ inputs.cpython_release }}.tgz ./tmp/
+          cd tmp/
+          tar xvf Python-${{ inputs.cpython_release }}.tgz
+          cd Python-${{ inputs.cpython_release }}
+
+          ./configure --prefix=$(realpath '../installation/')
+          make -j
+          make install -j
+
+          cd ../installation
+          ./bin/python3 -m test -uall
+
+      - name: "Upload the source artifacts"
+        uses: actions/upload-artifact@a8a3f3ad30e3422c9c7b888a15615d19a852ae32 # v3.1.3
+        with:
+          name: source
+          path: |
+            cpython/${{ inputs.cpython_release }}/src
+
+      - name: "Upload the docs artifacts"
+        uses: actions/upload-artifact@a8a3f3ad30e3422c9c7b888a15615d19a852ae32 # v3.1.3
+        # Conditionally run this step if there is a 'docs/' directory.
+        # Docs aren't built for alpha or beta releases.
+        if: ${{ hashFiles('cpython/${{ inputs.cpython_release }}/docs') != '' }}
```

Comment:
This conditional doesn't match the above, but I figure if docs got built then we'd want to upload them as artifacts?

Reply:
In this run:
https://github.com/python/release-tools/actions/runs/8350750234/job/22857827913

the docs upload was skipped for whatever reason. They were built though. The source artifacts uploaded just fine. I don't immediately see why that would be so I just built the docs locally after for this release. I compared the logs from my local run with the docs build on GHA, they look the same. They both have the line:

```
2024-03-19T22:28:44.9625823Z chdir'ing to Python-3.10.14/Doc
2024-03-19T22:28:44.9626238Z Removing doc build artifacts
```

This comes from here: https://github.com/python/release-tools/blob/c1fceb72228302f5d20956fb415677b37047694a/release.py#L342-L350

which you can see is immediately preceded by a `shutil.copytree`, which in my case at least copied everything where expected. So maybe `hashFiles` does something weird. IDK.